File: hackalem/resilience.py
"""Validate the optional C++ structural experiment without changing its values."""
from __future__ import annotations

import math


def validate_resilience(value, payload):
    """Check the v1 contract, including JSON types, ordering and node coverage."""
    def require(condition, message):
        if not condition:
            raise ValueError(f'resilience: {message}')

    require(isinstance(value, dict), 'must be an object')
    for key, expected in (('schema_version', '1.0'), ('connectivity', 'weak'),
                          ('scope', 'observed_graph')):
        require(value.get(key) == expected, f'{key} must be {expected}')
    scenarios = value.get('scenarios')
    require(isinstance(scenarios, list) and len(scenarios) == 8,
            'scenarios must contain all eight priority/volume experiments')
    expected_order = [('baseline', 0)] + [(strategy, k)
        for strategy in ('priority', 'volume') for k in (1, 3, 5, 10)]
    gids = {node['gid'] for node in payload['nodes']}
    for scenario, (strategy, k) in zip([value.get('baseline'), *scenarios], expected_order):
        label = f'{strategy}/{k}'
        require(isinstance(scenario, dict), f'{label} must be an object')
        require(scenario.get('strategy') == strategy, f'{label}: unexpected strategy/order')
        require(type(scenario.get('requested_k')) is int and scenario['requested_k'] == k,
                f'{label}: unexpected requested_k')
        removed = scenario.get('removed_gids')
        require(isinstance(removed, list) and all(isinstance(g, str) for g in removed),
                f'{label}: removed_gids must be an array of strings')
        require(len(removed) == len(set(removed)) == min(k, len(gids))
                and set(removed) <= gids, f'{label}: invalid removed_gids coverage')
        for key in ('remaining_nodes', 'remaining_edges', 'weak_components',
                    'isolated_nodes', 'largest_component_nodes'):
            require(type(scenario.get(key)) is int and scenario[key] >= 0,
                    f'{label}: {key} must be a nonnegative integer')
        for key in ('largest_component_share_remaining', 'removed_edge_sum_kzt',
                    'removed_edge_sum_share'):
            number = scenario.get(key)
            require(type(number) in (int, float) and math.isfinite(number) and number >= 0,
                    f'{label}: {key} must be a finite nonnegative number')
            if key != 'removed_edge_sum_kzt':
                require(number <= 1, f'{label}: {key} must be in [0,1]')
        remaining = scenario['remaining_nodes']
        require(remaining == len(gids) - len(removed), f'{label}: remaining_nodes differs')
        require(scenario['remaining_edges'] <= len(payload['edges']),
                f'{label}: remaining_edges exceeds input')
        for key in ('weak_components', 'isolated_nodes', 'largest_component_nodes'):
            require(scenario[key] <= remaining, f'{label}: {key} exceeds remaining_nodes')
        if remaining:
            require(scenario['weak_components'] > 0 and scenario['largest_component_nodes'] > 0,
                    f'{label}: nonempty graph must contain a component')
        else:
            require(scenario['remaining_edges'] == 0, f'{label}: empty graph contains edges')
        if strategy == 'baseline':
            require(scenario['remaining_edges'] == len(payload['edges'])
                    and scenario['removed_edge_sum_kzt'] == 0
                    and scenario['removed_edge_sum_share'] == 0,
                    'baseline must retain every edge and remove no turnover')

File: frontend/src/resilience.ts
export type ResilienceStrategy = 'priority' | 'volume';
export type ResilienceK = 1 | 3 | 5 | 10;
export const resilienceCounts: readonly ResilienceK[] = [1, 3, 5, 10];
export interface Scenario {
  strategy: 'baseline' | ResilienceStrategy;
  requested_k: number;
  removed_gids: string[];
  remaining_nodes: number;
  remaining_edges: number;
  weak_components: number;
  isolated_nodes: number;
  largest_component_nodes: number;
  largest_component_share_remaining: number;
  removed_edge_sum_kzt: number;
  removed_edge_sum_share: number;
}
export interface Resilience {
  schema_version: '1.0';
  connectivity: 'weak';
  scope: 'observed_graph';
  baseline: Scenario;
  scenarios: Scenario[];
}

const fail = (path: string, reason: string): never => {
  throw new Error(`Неверный формат «Устойчивость наблюдаемой сети» (${path}): ${reason}.`);
};
function object(value: unknown, path: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return fail(path, 'ожидается объект');
  return value as Record<string, unknown>;
}
function integer(value: unknown, path: string): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) return fail(path, 'ожидается целое неотрицательное число');
  return value;
}
function finite(value: unknown, path: string, share = false): number {
  if (typeof value !== 'number' || !Number.isFinite(value) || value < 0 || (share && value > 1)) {
    return fail(path, share ? 'ожидается конечная доля от 0 до 1' : 'ожидается конечное неотрицательное число');
  }
  return value;
}
function parseScenario(raw: unknown, path: string, nodeIds: ReadonlySet<string>, edgeCount: number): Scenario {
  const value = object(raw, path);
  if (value.strategy !== 'baseline' && value.strategy !== 'priority' && value.strategy !== 'volume') return fail(`${path}.strategy`, 'неизвестный способ отбора');
  const requested_k = integer(value.requested_k, `${path}.requested_k`);
  if (!Array.isArray(value.removed_gids)) return fail(`${path}.removed_gids`, 'ожидается массив строк gid');
  const removed_gids: string[] = [], seen = new Set<string>();
  for (const gid of value.removed_gids) {
    if (typeof gid !== 'string' || !nodeIds.has(gid)) return fail(`${path}.removed_gids`, 'gid должен быть строкой существующего клиента');
    if (seen.has(gid)) return fail(`${path}.removed_gids`, 'клиент указан повторно');
    seen.add(gid); removed_gids.push(gid);
  }
  const scenario: Scenario = {
    strategy: value.strategy, requested_k, removed_gids,
    remaining_nodes: integer(value.remaining_nodes, `${path}.remaining_nodes`),
    remaining_edges: integer(value.remaining_edges, `${path}.remaining_edges`),
    weak_components: integer(value.weak_components, `${path}.weak_components`),
    isolated_nodes: integer(value.isolated_nodes, `${path}.isolated_nodes`),
    largest_component_nodes: integer(value.largest_component_nodes, `${path}.largest_component_nodes`),
    largest_component_share_remaining: finite(value.largest_component_share_remaining, `${path}.largest_component_share_remaining`, true),
    removed_edge_sum_kzt: finite(value.removed_edge_sum_kzt, `${path}.removed_edge_sum_kzt`),
    removed_edge_sum_share: finite(value.removed_edge_sum_share, `${path}.removed_edge_sum_share`, true),
  };
  if (removed_gids.length !== Math.min(requested_k, nodeIds.size)) return fail(path, 'число исключённых клиентов не соответствует N и размеру исходного графа');
  if (scenario.remaining_nodes !== nodeIds.size - removed_gids.length) return fail(path, 'число оставшихся клиентов не согласовано с исключёнными');
  if (scenario.remaining_edges > edgeCount) return fail(path, 'оставшихся связей больше, чем в исходном графе');
  const { remaining_nodes: nodes, largest_component_nodes: largest, weak_components: components, isolated_nodes: isolated } = scenario;
  if (nodes === 0) {
    if (scenario.remaining_edges !== 0 || components !== 0 || isolated !== 0 || largest !== 0) return fail(path, 'у пустой модели связи, компоненты и изолированные клиенты должны быть равны нулю');
  } else {
    if (largest < 1 || largest > nodes || components < 1 || components > nodes - largest + 1 || isolated > components) {
      return fail(path, 'размеры компонент и число изолированных клиентов не согласованы');
    }
    if ((largest === 1 && isolated !== nodes) || (largest > 1 && isolated > nodes - largest)) return fail(path, 'число изолированных клиентов не согласовано с крупнейшей компонентой');
  }
  const expectedShare = nodes === 0 ? 0 : largest / nodes;
  if (Math.abs(scenario.largest_component_share_remaining - expectedShare) > 1e-8) return fail(path, 'доля крупнейшей компоненты не согласована с числом оставшихся клиентов');
  return scenario;
}

/** Validate prepared results; never recompute graph connectivity or choose removed clients. */
export function parseResilience(raw: unknown, nodeIds: ReadonlySet<string>, edgeCount: number): Resilience | undefined {
  if (raw === undefined) return undefined;
  const value = object(raw, 'meta.resilience');
  if (value.schema_version !== '1.0') return fail('schema_version', 'ожидается версия 1.0');
  if (value.connectivity !== 'weak') return fail('connectivity', 'ожидается weak');
  if (value.scope !== 'observed_graph') return fail('scope', 'ожидается observed_graph');
  const baseline = parseScenario(value.baseline, 'baseline', nodeIds, edgeCount);
  if (baseline.strategy !== 'baseline' || baseline.requested_k !== 0 || baseline.removed_gids.length !== 0 || baseline.remaining_edges !== edgeCount || baseline.removed_edge_sum_kzt !== 0 || baseline.removed_edge_sum_share !== 0) {
    return fail('baseline', 'исходная модель должна описывать полный граф без исключений');
  }
  if (!Array.isArray(value.scenarios) || value.scenarios.length !== 8) return fail('scenarios', 'ожидаются все 8 сочетаний двух способов и N=1/3/5/10');
  const scenarios = value.scenarios.map((scenario, index) => parseScenario(scenario, `scenarios[${index}]`, nodeIds, edgeCount));
  const combinations = new Set<string>();
  for (const scenario of scenarios) {
    if (scenario.strategy === 'baseline' || !resilienceCounts.includes(scenario.requested_k as ResilienceK)) return fail('scenarios', 'допустимы только priority/volume и N=1/3/5/10');
    const key = `${scenario.strategy}:${scenario.requested_k}`;
    if (combinations.has(key)) return fail('scenarios', 'сочетание способа и N повторяется');
    combinations.add(key);
    if (scenario.largest_component_nodes > baseline.largest_component_nodes) return fail('scenarios', 'крупнейшая компонента после исключения не может вырасти');
    if (nodeIds.size === 0 && (scenario.removed_edge_sum_kzt !== 0 || scenario.removed_edge_sum_share !== 0)) return fail('scenarios', 'у пустого исходного графа нет исключённого оборота');
  }
  return { schema_version: '1.0', connectivity: 'weak', scope: 'observed_graph', baseline, scenarios };
}

export function resilienceScenario(resilience: Resilience, strategy: ResilienceStrategy, count: ResilienceK): Scenario {
  const scenario = resilience.scenarios.find(item => item.strategy === strategy && item.requested_k === count);
  if (!scenario) return fail('scenarios', 'отсутствует выбранное сочетание способа и N');
  return scenario;
}
